### ISSUE TITLE
[FEATURE] Refector the inner classfor Setting 

### ISSUE THREAD

**Issue by mloufra** [opensearch-project/opensearch-sdk-java#457]
### Is your feature request related to a problem?
Too many parser class which implement `Function<Setting, ?>` and `Writeable` was added in `Setting.java`.

### What solution would you like?
Create a separate class for these writeable parsers and store them in one place instead of `Setting.java` class

### What alternatives have you considered?
_A clear and concise description of any alternative solutions or features you've considered._

### Do you have any additional context?
_Add any other context or screenshots about the feature request here._

**Comment by mloufra** [comment]
> Can you add more details in this issue @mloufra? Add a link of the comment from the PR and you can also link the file here

sure

**Comment by dbwiddis** [comment]
Good discussion to have and I'll toss in a [link to a comment](https://github.com/opensearch-project/OpenSearch/pull/5597#discussion_r1054484606) by @reta when `WriteableSetting` was merged.

TLDR: We can refactor the existing code but most of it still needs to exist, it's just a question of how to reorganize the logic and where to put it, and I somewhat prefer the existing implementation @mloufra is doing to address #349. 

Extended background:

1. The `Setting<T>` class is not `Writeable` and includes up to three non-`Writable` arguments in its public constructors.  See for example [this constructor](https://opensearch.org/javadocs/2.5.0/OpenSearch/server/build/docs/javadoc/org/opensearch/common/settings/Setting.html#%3Cinit%3E(org.opensearch.common.settings.Setting.Key,java.util.function.Function,java.util.function.Function,org.opensearch.common.settings.Setting.Validator,org.opensearch.common.settings.Setting.Property...)), specifically the default value, parser, and validator:
   - Settings are internally stored as a string (writeable)
   - Default value can refer to another setting (potentially a long daisy chain) that all ends at a string
   - Parser turns the String into a type `T`, each of which has a different return value, and necessitates either returning the specific type (multiple parsers) or having multiple methods, each of which returns the specific type and the others of which are invalid, or returning a more general type (e.g., `java.lang.Number` that still has [type-specific getters](https://docs.oracle.com/javase/8/docs/api/java/lang/Number.html) (e.g., `intValue()`, `longValue()`, etc.)
   - Parsers also often perform numeric validation (min/max value)
   - Validator performs other validation, primarily for things the parser can't (e.g., String validation).

2. Ideally we would want to update the javadoc to specify that `T` should be writeable and handle it.  We're still going to have to have a `Parser` that handles it and returns its value.  
   - I currently favor the existing implementation because users can either write their own parser (e.g., `BigIntegerParser()` or maybe `IPAddressParser` or `JWTParser`, etc.) using the existing parsers as a pattern.
   - We could certainly move the parsers to a different class/classes but they currently leverage some private methods inside `Setting` that would need to be moved or made public to support.

3. The initial work I did in #4519 (feature branch, eventually merged into main in #5597) was the minimum necessary to get all the `Setting` usages in the AD plugin minimally working by transferring over the original value. 
   - Unfortunately this lost reference to the original parsing functions (e.g., non-negative numeric values).  In an attempt not to edit the actual `Setting` class, I created a new `WriteableSetting` class that attempted to centralize all the switch/case logic of handling the various types.  
   - A lot of that logic _could_ be moved inside each specific writeable parser class, and that seems like a reasonable thing to do, but we're still going to need a different parser class for each `T` we handle.

So, if I were refactoring it, I'd probably make new standalone classes for each of the different `T` classes we currently handle (Boolean, Float, Double, Byte, Short, Int, Long, TimeValue, ByteSizeValue) which are mostly the existing inner classes, but move in some of the (currently private) parsing functions from `Setting` and the writeTo/streaminput constructor logic from `WriteableSetting`, such that `WriteableSetting` can go away, `Setting` will be smaller, and we'll still keep the current functionality.
